packages/polling-controller/src/PollingController.ts

```diff
@@ -0,0 +1,120 @@
+import { BaseControllerV2 } from '@metamask/base-controller';
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import type { NetworkClientId } from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+import { v1 as random } from 'uuid';
+
+export type PollingCompleteType<N extends string> = {
+  type: `${N}:pollingComplete`;
+  payload: [string];
+};
+
+/**
+ * PollingController is an abstract class that implements the polling
+ * functionality for a controller. It is meant to be extended by a controller
+ * that needs to poll for data by networkClientId.
+ *
+ */
+export default abstract class PollingController<
+  Name extends string,
+  State extends Record<string, Json>,
+  messenger extends RestrictedControllerMessenger<
+    Name,
+    any,
+    PollingCompleteType<Name> | any,
+    string,
+    string
+  >,
+> extends BaseControllerV2<Name, State, messenger> {
+  readonly #intervalLength = 1000;
+
+  private readonly networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
+    new Map();
+
+  private readonly intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
+
+  /**
+   * Starts polling for a networkClientId
+   *
+   * @param networkClientId - The networkClientId to start polling for
+   * @returns void
+   */
+  start(networkClientId: NetworkClientId) {
```

Comment:
Should we support polling controllers which may only have 1 polling loop active per networkClientId? Perhaps we could have an option passed into the constructor to indicate this? for example, if the blocktracker was a controller, we would only want to have 1 active block polling per network.

Reply:
Good question. I can't think of any controllers where we'd want this offhand though.